syft/source/directory_resolver.go

```diff
@@ -1,41 +1,211 @@
+	}
+
+	r := directoryResolver{
+		path:          root,
+		cwd:           cwd,
+		fileTree:      filetree.NewFileTree(),
+		infos:         make(map[file.ID]os.FileInfo),
+		pathFilterFns: pathFilters,
+		errPaths:      make(map[string]error),
+	}
+
+	// why account for multiple roots? To cover cases when there is a symlink that references above the root path,
+	// in which case we need to additionally index where the link resolves to. it's for this reason why the filetree
+	// must be relative to the root of the filesystem (and not just relative to the given path).
+	roots := []string{root}
+loop:
+	for {
+		var p string
+		switch len(roots) {
+		case 0:
+			break loop
+		case 1:
+			p, roots = roots[0], nil
+		default:
+			p, roots = roots[0], roots[1:]
+		}
+
+		additionalRoots, err := r.indexTree(p)
+		if err != nil {
+			return nil, fmt.Errorf("unable to index filesystem: %w", err)
+		}
+		roots = append(roots, additionalRoots...)
+	}
+
+	return &r, nil
+}
+
+func (r *directoryResolver) indexTree(root string) ([]string, error) {
+	log.Infof("indexing filesystem path=%q", root)
+	var err error
+	root, err = filepath.Abs(root)
+	if err != nil {
+		return nil, err
+	}
+	var roots []string
+	stager, prog := indexingProgress(root)
+	defer prog.SetCompleted()
+
+	return roots, filepath.Walk(root,
+		func(p string, info os.FileInfo, err error) error {
+			stager.Current = p
+
+			// ignore any path which a filter function returns true
+			for _, fn := range r.pathFilterFns {
+				if fn(p) {
+					return nil
+				}
+			}
+
+			// permission denied, IO error, etc... we keep track of the paths we can't see, but continue with indexing
+			if err != nil {
+				log.Warnf("unable to index path=%q: %+v", p, err)
+				r.errPaths[p] = err
+				return nil
+			}
+
+			// link cycles could cause a revisit --we should not allow this
+			if r.fileTree.HasPath(file.Path(p)) {
+				return nil
+			}
+
+			newRoot, err := r.addPathToIndex(p, info)
+			if err != nil {
+				return err
+			}
+
+			if newRoot != "" {
+				roots = append(roots, newRoot)
+			}
+
+			return nil
+		})
+}
+
+func (r directoryResolver) addPathToIndex(p string, info os.FileInfo) (string, error) {
+	var ref *file.Reference
+	var err error
+	var newRoot string
+
+	switch newFileTypeFromMode(info.Mode()) {
+	case SymbolicLink:
+		linkTarget, err := os.Readlink(p)
+		if err != nil {
+			if errors.Is(err, os.ErrPermission) {
+				// don't allow for permission errors to stop indexing, keep track of the paths and continue.
+				log.Warnf("unable to index symlink=%q: %+v", p, err)
+				r.errPaths[p] = err
+				return "", nil
+			}
+			return "", fmt.Errorf("unable to readlink for path=%q: %+v", p, err)
+		}
+		ref, err = r.fileTree.AddSymLink(file.Path(p), file.Path(linkTarget))
+		if err != nil {
+			return "", err
+		}
+
+		targetAbsPath := linkTarget
+		if !filepath.IsAbs(targetAbsPath) {
+			targetAbsPath = filepath.Clean(filepath.Join(path.Dir(p), linkTarget))
```

Comment:
why is the clean function call necessary here for absolute paths?

Reply:
also would it be necessary on line 184?

Reply:
the linkTarget is not guaranteed to be an absolute path. For instance:
```
filepath.Clean(filepath.Join(path.Dir("/some/abs/path"),"../../relative/path"))
```

re: 184, we don't know if `r.path` is relative or absolute, however, we do know by that point that `userPath` is absolute from the line before --so it shouldn't be possible to reference anything outside of the tree's root path (`r.path`).